### DISCUSSION TITLE
flexric's slicing xapp has an error

### DISCUSSION THREAD

**Discussion by joshuamoorexyz** [srsran/srsRAN_Project#218] [General Help]
When trying to run the python xapp_slice_moni_ctrl after having sucessfully run the kpimoni xapp from the instructions in srsrans xapp section there is some issue.

The xapps logs:
joshuamoorexyz@joshuamoorexyz-B550-UD-AC:~/Github/flexric/build$ python3 examples/xApp/python3/xapp_slice_moni_ctrl.py 
Setting the config -c file to /usr/local/etc/flexric/flexric.conf
Setting path -p for the shared libraries to /usr/local/lib/flexric/
[xAapp]: Initializing ... 
[xApp]: nearRT-RIC IP Address = 127.0.0.1, PORT = 36422
[E2 AGENT]: Opening plugin from path = /usr/local/lib/flexric/librlc_sm.so 
[E2 AGENT]: Opening plugin from path = /usr/local/lib/flexric/libpdcp_sm.so 
[E2 AGENT]: Opening plugin from path = /usr/local/lib/flexric/libslice_sm.so 
[E2 AGENT]: Opening plugin from path = /usr/local/lib/flexric/libtc_sm.so 
[E2 AGENT]: Opening plugin from path = /usr/local/lib/flexric/libkpm_sm.so 
[E2 AGENT]: Opening plugin from path = /usr/local/lib/flexric/libmac_sm.so 
[E2 AGENT]: Opening plugin from path = /usr/local/lib/flexric/libgtp_sm.so 
[NEAR-RIC]: Loading SM ID = 143 with def = RLC_STATS_V0 
[NEAR-RIC]: Loading SM ID = 144 with def = PDCP_STATS_V0 
[NEAR-RIC]: Loading SM ID = 145 with def = SLICE_STATS_V0 
[NEAR-RIC]: Loading SM ID = 146 with def = TC_STATS_V0 
[NEAR-RIC]: Loading SM ID = 147 with def = ORAN-E2SM-KPM 
[NEAR-RIC]: Loading SM ID = 142 with def = MAC_STATS_V0 
[NEAR-RIC]: Loading SM ID = 148 with def = GTP_STATS_V0 
Filename = /tmp/xapp_db_1695234449766823 
 [xApp]: E42 SETUP-REQUEST sent
adding event fd = 8 ev-> 4 
[xApp]: E42 SETUP-RESPONSE received
[xApp]: xApp ID = 7 
Registered E2 Nodes = 1 
Pending event size before remove = 1 
Generated of req_id = 1 
[xApp]: RIC SUBSCRIPTION REQUEST sent
adding event fd = 8 ev-> 5 
python3: /home/joshuamoorexyz/Github/flexric/src/xApp/e42_xapp.c:252: e2_event_loop_xapp: Assertion `*e.p_ev != E42_RIC_SUBSCRIPTION_REQUEST_PENDING_EVENT && "Timeout waiting for Report. Connection lost with the RIC?"' failed.
Aborted (core dumped)

The nearRT RIC Logs:

joshuamoorexyz@joshuamoorexyz-B550-UD-AC:~/Github/flexric/build/examples/ric$ sudo ./nearRT-RIC 
Setting the config -c file to /usr/local/etc/flexric/flexric.conf
Setting path -p for the shared libraries to /usr/local/lib/flexric/
[NEAR-RIC]: nearRT-RIC IP Address = 127.0.0.1, PORT = 36421
[NEAR-RIC]: Initializing 
[NEAR-RIC]: Loading SM ID = 143 with def = RLC_STATS_V0 
[NEAR-RIC]: Loading SM ID = 144 with def = PDCP_STATS_V0 
[NEAR-RIC]: Loading SM ID = 145 with def = SLICE_STATS_V0 
[NEAR-RIC]: Loading SM ID = 146 with def = TC_STATS_V0 
[NEAR-RIC]: Loading SM ID = 147 with def = ORAN-E2SM-KPM 
[NEAR-RIC]: Loading SM ID = 142 with def = MAC_STATS_V0 
[NEAR-RIC]: Loading SM ID = 148 with def = GTP_STATS_V0 
[iApp]: Initializing ... 
[iApp]: nearRT-RIC IP Address = 127.0.0.1, PORT = 36422
fd created with 6 
Received message with id = 411, port = 13952 
[E2AP] Received SETUP-REQUEST from PLMN   1. 1 Node ID 411 RAN type ngran_gNB
[NEAR-RIC]: Accepting RAN function ID 147 with def = `0ORAN-E2SM-KPM 
[NEAR-RIC]: Accepting interfaceType 0
[iApp]: E42 SETUP-REQUEST received
[iApp]: E42 SETUP-RESPONSE sent
[iApp]: SUBSCRIPTION-REQUEST xapp_ric_id->ric_id.ran_func_id 145  
[E2AP] SUBSCRIPTION REQUEST generated
[NEAR-RIC]: nb_id 411 port = 13952  
nearRT-RIC: /home/joshuamoorexyz/Github/flexric/src/ric/msg_handler_ric.c:204: e2ap_handle_subscription_failure_ric: Assertion `0!=0 && "Not implemented"' failed.
Aborted

A video showing the error happening: 

https://www.youtube.com/watch?v=2_C2CDiv148

I understand that if it is simply not supported yet it may not be possible without some effort on my part, if that is the case can you give some guidelines on how i could approach solving this.

**Comment by yagoda** [comment]
The subscription was rejected by the e2 agent in the gnb, filter the gnb
logs with "E2" and share the output.

**Reply by joshuamoorexyz** [reply]
This seems the most relevant. if not i can share the whole log 

2023-09-20T18:52:15.997549 [E2-ASN1-PCK] [I] Received PDU of 40 bytes
2023-09-20T18:52:15.997669 [E2      ] [I] Handling E2 PDU of type initiatingMessage
2023-09-20T18:52:15.997673 [E2      ] [I] E2AP SDU, "initiatingMessage.RICsubscriptionRequest"
2023-09-20T18:52:15.997674 [E2      ] [I] Received RIC Subscription Request
2023-09-20T18:52:15.997676 [E2      ] [I] E2AP: Received subscription request
2023-09-20T18:52:15.997679 [E2-SUBSCRIBER] [E] RAN function ID=145 not supported
2023-09-20T18:52:15.997720 [E2      ] [I] E2AP: Sending subscription failure
2023-09-20T18:52:16.002815 [UL-PHY0 ] [I] [   481.9] PUCCH: rnti=0x4601 format=1 prb1=51

**Comment by yagoda** [comment]
There you have it.
function ID=145 not supported

The xapp/ric should read the setup request and only use function IDs the
gnb provides, unfortunately its hardcoded in some of the xapps it seems,
you just need to change that in the xapp

**Reply by joshuamoorexyz** [reply]
Thanks i will try that and see if the xapp functions properly.
